File: web/src/lib/api/schemas/tables.ts
import { z } from "zod";

export const tableDataTypeSchema = z.enum([
  "text",
  "integer",
  "float",
  "boolean",
  "date",
  "timestamp",
  "json",
]);

export const sortDirectionSchema = z.enum(["asc", "desc"]);

export const filterOperatorSchema = z.enum([
  "eq",
  "neq",
  "gt",
  "gte",
  "lt",
  "lte",
  "in",
  "contains",
  "starts_with",
  "ends_with",
  "is_null",
  "not_null",
]);

export const aggregateFunctionSchema = z.enum([
  "count",
  "sum",
  "avg",
  "min",
  "max",
]);

export const importStatusSchema = z.enum([
  "pending",
  "running",
  "completed",
  "failed",
]);

export const exportFormatSchema = z.enum(["csv", "json", "xlsx"]);

export const sourceActorSchema = z.enum(["user", "agent", "import"]);

export const importFormatSchema = z.enum(["csv", "json", "xlsx"]);

export const referenceTableColumnInputSchema = z.object({
  name: z.string().min(1).max(63),
  data_type: tableDataTypeSchema,
  nullable: z.boolean().default(true),
  description: z.string().max(1024).nullable().optional(),
  semantic_hint: z.string().max(128).nullable().optional(),
  constraints_json: z.record(z.string(), z.unknown()).nullable().optional(),
  default_value: z.unknown().nullable().optional(),
});

export const referenceTableCreateInputSchema = z.object({
  name: z.string().min(1).max(63),
  title: z.string().max(255).nullable().optional(),
  description: z.string().max(2048).nullable().optional(),
  columns: z.array(referenceTableColumnInputSchema).min(1),
});

export const referenceTableUpdateInputSchema = z.object({
  title: z.string().max(255).nullable().optional(),
  description: z.string().max(2048).nullable().optional(),
  columns: z.array(referenceTableColumnInputSchema).nullable().optional(),
});

export const referenceTableColumnSchema = z.object({
  id: z.string(),
  name: z.string(),
  position: z.number().int().nonnegative(),
  data_type: tableDataTypeSchema,
  nullable: z.boolean(),
  description: z.string().nullable().optional(),
  semantic_hint: z.string().nullable().optional(),
  constraints_json: z.record(z.string(), z.unknown()).nullable().optional(),
  default_value: z.unknown().nullable().optional(),
});

export const referenceTableSummarySchema = z.object({
  id: z.string(),
  name: z.string(),
  title: z.string().nullable().optional(),
  description: z.string().nullable().optional(),
  row_count: z.number().int().nonnegative(),
  created_at: z.string(),
  updated_at: z.string(),
});

export const referenceTableDetailSchema = referenceTableSummarySchema.extend({
  columns: z.array(referenceTableColumnSchema),
});

export const queryFilterSchema = z.object({
  field: z.string().min(1).max(63),
  op: filterOperatorSchema,
  value: z.unknown().nullable().optional(),
});

export const querySortSchema = z.object({
  field: z.string().min(1).max(63),
  direction: sortDirectionSchema.default("asc"),
});

export const queryAggregateSchema = z.object({
  function: aggregateFunctionSchema,
  field: z.string().max(63).nullable().optional(),
  alias: z.string().max(63).nullable().optional(),
});

export const rowsQueryInputSchema = z.object({
  filters: z.array(queryFilterSchema).default([]),
  sorts: z.array(querySortSchema).default([]),
  page: z.number().int().min(1).default(1),
  page_size: z.number().int().min(1).max(500).default(50),
  group_by: z.array(z.string()).default([]),
  aggregates: z.array(queryAggregateSchema).default([]),
});

export const queriedRowSchema = z.object({
  id: z.string(),
  version: z.number().int().nonnegative(),
  values_json: z.record(z.string(), z.unknown()),
  created_at: z.string(),
  updated_at: z.string(),
});

export const rowsQueryResponseSchema = z.object({
  total_rows: z.number().int().nonnegative(),
  page: z.number().int().min(1),
  page_size: z.number().int().min(1),
  rows: z.array(queriedRowSchema),
  aggregate_row: z.record(z.string(), z.unknown()).nullable().optional(),
  grouped_rows: z.array(z.record(z.string(), z.unknown())).default([]),
  provenance: z.record(z.string(), z.unknown()),
});

export const rowUpsertSchema = z.object({
  row_id: z.string().nullable().optional(),
  values_json: z.record(z.string(), z.unknown()),
  source_actor: sourceActorSchema.default("user"),
  source_ref: z.string().max(255).nullable().optional(),
});

export const rowsBatchInputSchema = z
  .object({
    upserts: z.array(rowUpsertSchema).default([]),
    delete_row_ids: z.array(z.string()).default([]),
  })
  .superRefine((value, ctx) => {
    if (value.upserts.length === 0 && value.delete_row_ids.length === 0) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: "Provide at least one upsert or delete operation.",
        path: ["upserts"],
      });
    }
  });

export const rowsBatchResultSchema = z.object({
  inserted: z.number().int().nonnegative(),
  updated: z.number().int().nonnegative(),
  deleted: z.number().int().nonnegative(),
});

export const importStartInputSchema = z.object({
  attachment_id: z.string(),
  source_format: importFormatSchema.nullable().optional(),
  has_header: z.boolean().default(true),
  delimiter: z.string().nullable().optional(),
  column_overrides: z.record(z.string(), tableDataTypeSchema).default({}),
});

export const importJobSummarySchema = z.object({
  id: z.string(),
  table_id: z.string(),
  status: importStatusSchema,
  source_filename: z.string().nullable().optional(),
  source_format: importFormatSchema.nullable().optional(),
  total_rows: z.number().int().nonnegative(),
  inserted_rows: z.number().int().nonnegative(),
  updated_rows: z.number().int().nonnegative(),
  deleted_rows: z.number().int().nonnegative(),
  error_count: z.number().int().nonnegative(),
  errors_json: z.array(z.record(z.string(), z.unknown())),
  created_at: z.string(),
  started_at: z.string().nullable().optional(),
  finished_at: z.string().nullable().optional(),
  inferred_columns: z
    .array(z.record(z.string(), z.unknown()))
    .nullable()
    .optional(),
  provenance: z.record(z.string(), z.unknown()).nullable().optional(),
});

export const exportInputSchema = z.object({
  format: exportFormatSchema,
  query: rowsQueryInputSchema.nullable().optional(),
  include_header: z.boolean().default(true),
});

export type TableDataType = z.infer<typeof tableDataTypeSchema>;
export type SortDirection = z.infer<typeof sortDirectionSchema>;
export type FilterOperator = z.infer<typeof filterOperatorSchema>;
export type AggregateFunction = z.infer<typeof aggregateFunctionSchema>;
export type ImportStatus = z.infer<typeof importStatusSchema>;
export type ExportFormat = z.infer<typeof exportFormatSchema>;
export type ImportFormat = z.infer<typeof importFormatSchema>;

export type ReferenceTableColumnInput = z.infer<
  typeof referenceTableColumnInputSchema
>;
export type ReferenceTableCreateInput = z.infer<
  typeof referenceTableCreateInputSchema
>;
export type ReferenceTableUpdateInput = z.infer<
  typeof referenceTableUpdateInputSchema
>;
export type ReferenceTableColumn = z.infer<typeof referenceTableColumnSchema>;
export type ReferenceTableSummary = z.infer<typeof referenceTableSummarySchema>;
export type ReferenceTableDetail = z.infer<typeof referenceTableDetailSchema>;
export type QueryFilter = z.infer<typeof queryFilterSchema>;
export type QuerySort = z.infer<typeof querySortSchema>;
export type QueryAggregate = z.infer<typeof queryAggregateSchema>;
export type RowsQueryInput = z.infer<typeof rowsQueryInputSchema>;
export type QueriedRow = z.infer<typeof queriedRowSchema>;
export type RowsQueryResponse = z.infer<typeof rowsQueryResponseSchema>;
export type RowUpsert = z.infer<typeof rowUpsertSchema>;
export type RowsBatchInput = z.infer<typeof rowsBatchInputSchema>;
export type RowsBatchResult = z.infer<typeof rowsBatchResultSchema>;
export type ImportStartInput = z.infer<typeof importStartInputSchema>;
export type ImportJobSummary = z.infer<typeof importJobSummarySchema>;
export type ExportInput = z.infer<typeof exportInputSchema>;

File: server/db/models/tables.py
from __future__ import annotations

from datetime import datetime
from uuid import UUID, uuid4

from sqlalchemy import (
    Boolean,
    DateTime,
    ForeignKey,
    Index,
    Integer,
    JSON,
    String,
    Text,
    UniqueConstraint,
    func,
)
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column, relationship

from server.db.base import Base


class ReferenceTable(Base):
    __tablename__ = "reference_tables"

    id: Mapped[UUID] = mapped_column(primary_key=True, default=uuid4)
    name: Mapped[str] = mapped_column(String(63), nullable=False, unique=True, index=True)
    title: Mapped[str | None] = mapped_column(String(255), nullable=True)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    row_count: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
        default=0,
        server_default="0",
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    columns: Mapped[list[ReferenceTableColumn]] = relationship(
        "ReferenceTableColumn",
        back_populates="table",
        cascade="all, delete-orphan",
        order_by="ReferenceTableColumn.position",
    )
    rows: Mapped[list[ReferenceTableRow]] = relationship(
        "ReferenceTableRow",
        back_populates="table",
        cascade="all, delete-orphan",
        order_by=lambda: (ReferenceTableRow.created_at, ReferenceTableRow.id),
    )
    import_jobs: Mapped[list[ReferenceTableImportJob]] = relationship(
        "ReferenceTableImportJob",
        back_populates="table",
        cascade="all, delete-orphan",
        order_by=lambda: (ReferenceTableImportJob.created_at, ReferenceTableImportJob.id),
    )


class ReferenceTableColumn(Base):
    __tablename__ = "reference_table_columns"
    __table_args__ = (
        UniqueConstraint("table_id", "name", name="uq_reference_table_columns_table_name"),
        UniqueConstraint("table_id", "position", name="uq_reference_table_columns_table_position"),
        Index(
            "ix_reference_table_columns_table_position",
            "table_id",
            "position",
        ),
    )

    id: Mapped[UUID] = mapped_column(primary_key=True, default=uuid4)
    table_id: Mapped[UUID] = mapped_column(
        ForeignKey("reference_tables.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    name: Mapped[str] = mapped_column(String(63), nullable=False)
    position: Mapped[int] = mapped_column(Integer, nullable=False)
    data_type: Mapped[str] = mapped_column(String(32), nullable=False)
    nullable: Mapped[bool] = mapped_column(
        Boolean,
        nullable=False,
        default=True,
        server_default="true",
    )
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    semantic_hint: Mapped[str | None] = mapped_column(String(128), nullable=True)
    constraints_json: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    default_json: Mapped[dict | int | float | str | bool | list | None] = mapped_column(
        JSON,
        nullable=True,
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    table: Mapped[ReferenceTable] = relationship("ReferenceTable", back_populates="columns")


class ReferenceTableImportJob(Base):
    __tablename__ = "reference_table_import_jobs"

    id: Mapped[UUID] = mapped_column(primary_key=True, default=uuid4)
    table_id: Mapped[UUID] = mapped_column(
        ForeignKey("reference_tables.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    attachment_id: Mapped[UUID | None] = mapped_column(
        ForeignKey("attachments.id", ondelete="SET NULL"),
        nullable=True,
    )
    status: Mapped[str] = mapped_column(
        String(32),
        nullable=False,
        default="pending",
        server_default="pending",
    )
    source_filename: Mapped[str | None] = mapped_column(String(512), nullable=True)
    source_format: Mapped[str | None] = mapped_column(String(16), nullable=True)
    started_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    finished_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    total_rows: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
        default=0,
        server_default="0",
    )
    inserted_rows: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
        default=0,
        server_default="0",
    )
    updated_rows: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
        default=0,
        server_default="0",
    )
    deleted_rows: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
        default=0,
        server_default="0",
    )
    error_count: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
        default=0,
        server_default="0",
    )
    errors_json: Mapped[list[dict]] = mapped_column(
        JSON,
        nullable=False,
        default=list,
    )
    created_by: Mapped[str | None] = mapped_column(String(128), nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    table: Mapped[ReferenceTable] = relationship("ReferenceTable", back_populates="import_jobs")
    attachment: Mapped[Attachment | None] = relationship("Attachment")
    rows: Mapped[list[ReferenceTableRow]] = relationship(
        "ReferenceTableRow",
        back_populates="import_job",
        order_by=lambda: (ReferenceTableRow.created_at, ReferenceTableRow.id),
    )


class ReferenceTableRow(Base):
    __tablename__ = "reference_table_rows"
    __table_args__ = (
        Index("ix_reference_table_rows_table_updated", "table_id", "updated_at"),
        Index("ix_reference_table_rows_table_import_job", "table_id", "import_job_id"),
    )

    id: Mapped[UUID] = mapped_column(primary_key=True, default=uuid4)
    table_id: Mapped[UUID] = mapped_column(
        ForeignKey("reference_tables.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    import_job_id: Mapped[UUID | None] = mapped_column(
        ForeignKey("reference_table_import_jobs.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    version: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
        default=1,
        server_default="1",
    )
    values_json: Mapped[dict] = mapped_column(JSONB, nullable=False)
    source_actor: Mapped[str] = mapped_column(
        String(16),
        nullable=False,
        default="user",
        server_default="user",
    )
    source_ref: Mapped[str | None] = mapped_column(String(255), nullable=True)
    created_by: Mapped[str | None] = mapped_column(String(128), nullable=True)
    updated_by: Mapped[str | None] = mapped_column(String(128), nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    table: Mapped[ReferenceTable] = relationship("ReferenceTable", back_populates="rows")
    import_job: Mapped[ReferenceTableImportJob | None] = relationship(
        "ReferenceTableImportJob",
        back_populates="rows",
    )
